juju/juju
# migration and upgrade steps for portsDoc subnetID from a CIDR to actual ID.

# Description of change

Migration and upgrade steps for portsDoc subnetID from a CIDR to actual ID.  If we're upgrading or migrating from a version of juju where the portsDoc.SubnetID is a CIDR, all CIDRs must be unique.  So there will be no issue finding the matching subnet.ID for that CIDR.

Placement of these steps is important.  The migration step depends on the "recreate subnets with IDs" step being run first.  The upgrade step depends on subnets being upgraded before machines. 

Fix migration bug found during testing.
## QA steps

Use an OpenStack cloud or another which use the firewall rules.

Test migration:
1. juju bootstrap <openstack> move-to
2. juju bootstrap <openstack> tuesday
3. juju add-model move-me
4. juju deploy mysql ; juju expose mysql
3. juju-db.bash ; db.openedPorts.find()
3. wait for deployment to finish
3. juju migrate move-me move-to
3. verify migration and no related errors in logs
3. juju switch move-me
3. juju-db.bash ; db.openedPorts.find()

Test upgrade:
1. sudo snap refresh juju --channel 2.6/edge
2. /snap/bin/juju bootstrap <openstack> tuesday
4. /snap/bin/juju deploy mysql ; juju expose mysql
3. wait for deployment to finish
3. juju upgrade-controller --build-agent
3. juju upgrade-juju
3. verify upgrade and no related errors in logs.
3. juju-db.bash ; db.openedPorts.find()
